src/foam/nanos/column/NestedPropertiesExpression.js

```diff
@@ -81,61 +81,120 @@ foam.CLASS({
         {
           name: 'i',
           class: 'Int'
+        },
+        {
+          name: 'expr',
+          javaType: 'foam.mlang.Expr'
         }
       ],
-      code: function (of, propName, i) {
-        if ( i === propName.length - 1 )
-          return of.getAxiomByName(propName[i]);
+      code: function (of, propName, i, expr) {
         var prop = of.getAxiomByName(propName[i]);
-        return foam.mlang.Expressions.create().DOT(prop, this.returnPropExpr(prop, this.returnDotExprForNestedProperty(prop.of, propName, ++i)));
+        if ( ! prop ) return null;
+
+        var propExpr = this.buildPropertyExpr(prop, expr);
+
+        if ( i === propName.length - 1 )
+          return propExpr;
+
+        return this.returnDotExprForNestedProperty(prop.of, propName, ++i, propExpr);
       },
       javaCode: `
         ClassInfo ci = of;
-        PropertyInfo p = (PropertyInfo) ci.getAxiomByName(propName[i]);
-        if ( i == propName.length - 1 ) {
-          return p;
-        }
+        PropertyInfo prop = (PropertyInfo) ci.getAxiomByName(propName[i]);
+
+        if ( prop == null ) return null;
+
+        Boolean isPropAReference = isPropertyAReference(prop);
+        expr = buildPropertyExpr(prop, expr);
+
+        if ( i == propName.length - 1 )
+          return expr;
+        
+        ci = getPropertyValueClass(ci, prop);
+        if ( ci == null )
+          return null; 
         
-        StringBuilder sb = new StringBuilder("find");
+        return returnDotExprForNestedProperty(ci, propName, ++i, expr);
+      `
+    },
+    {
+      name: 'getPropertyValueClass',
+      javaType: 'ClassInfo',
+      args: [
+        {
+          name: 'ci',
+          type: 'Object',
+          javaType: 'foam.core.ClassInfo'
+        },
+        {
+          name: 'prop',
+          javaType: 'foam.core.PropertyInfo',
+        }
+      ],
+      javaCode: `
         try {
+          Boolean isPropAReference = isPropertyAReference(prop);
```

Comment:
isPropertyAReference does a reflection call and this method does the same reflection call again on line 141. If we have something like `getFinderMethod(prop)` then reuse the result there would be one reflection call.

Reply:
fixed